test/integ/sso/test_unit_mfa_cache.py

```diff
@@ -0,0 +1,196 @@
+        elif mock_post_req_cnt == 2:
+            # check associated mfa token and issue a new mfa token
+            # note: Normally, backend doesn't issue a new mfa token in this case, we do it here only to test
+            # whether the driver can replace the old token when server provides a new token
+            assert body['data']['SESSION_PARAMETERS'].get('CLIENT_REQUEST_MFA_TOKEN') is True
+            assert body['data']['TOKEN'] == 'MFA_TOKEN'
+            ret = {
+                'success': True,
+                'message': None,
+                'data': {
+                    'token': 'NEW_TOKEN',
+                    'masterToken': 'NEW_MASTER_TOKEN',
+                    'mfaToken': 'NEW_MFA_TOKEN'
+                }}
+        elif mock_post_req_cnt == 4:
+            # check new mfa token
+            assert body['data']['SESSION_PARAMETERS'].get('CLIENT_REQUEST_MFA_TOKEN') is True
+            assert body['data']['TOKEN'] == 'NEW_MFA_TOKEN'
+            ret = {
+                'success': True,
+                'message': None,
+                'data': {
+                    'token': 'NEW_TOKEN',
+                    'masterToken': 'NEW_MASTER_TOKEN',
+                }}
+        elif mock_post_req_cnt == 6:
+            # mock a failed log in
+            ret = {
+                'success': False,
+                'message': None,
+                'data': {}
+            }
+        elif mock_post_req_cnt == 7:
+            assert body['data']['SESSION_PARAMETERS'].get('CLIENT_REQUEST_MFA_TOKEN') is True
+            assert 'TOKEN' not in body['data']
+            ret = {
+                'success': True,
+                'data': {
+                    'token': 'TOKEN',
+                    'masterToken': 'MASTER_TOKEN'
+                }
+            }
+        elif mock_post_req_cnt in [1, 3, 5, 8]:
+            # connection.close()
+            ret = {'success': True}
+        mock_post_req_cnt += 1
+        return ret
+
+    def mock_del_password(system, user):
+        LOCAL_CACHE.pop(system + user, None)
+
+    def mock_set_password(system, user, pwd):
+        LOCAL_CACHE[system + user] = pwd
+
+    def mock_get_password(system, user):
+        return LOCAL_CACHE.get(system + user, None)
+
+    global mock_post_req_cnt
+    mock_post_req_cnt = 0
+
+    # POST requests mock
+    mockSnowflakeRestfulPostRequest.side_effect = mock_post_request
+
+    def test_body():
+        account = 'testaccount'
+        user = 'testuser'
+        pwd = 'testpwd'
+        authenticator = 'username_password_mfa'
+        host = 'testaccount.snowflakecomputing.com'
+
+        delete_temporary_credential(host=host, user=user, cred_type=MFA_TOKEN)
+
+        # first connection, no mfa token cache
+        con = snowflake.connector.connect(
+            account=account,
+            user=user,
+            password=pwd,
+            host=host,
+            authenticator=authenticator,
+            client_request_mfa_token=True,
+        )
```

Comment:
I think you could have achieved the same thing with a little less typing. But I'm only pointing this out since you said you wanted to learn a more Pythonic coding style.
```suggestion
        conn_cfg = {
            'account': 'testaccount'
            'user': 'testuser'
            'password': 'testpwd'
            'authenticator': 'username_password_mfa'
            'host': 'testaccount.snowflakecomputing.com'
        }

        delete_temporary_credential(host=host, user=user, cred_type=MFA_TOKEN)

        # first connection, no mfa token cache
        con = snowflake.connector.connect(
            client_request_mfa_token=True,
            **conn_cfg
        )
```

Reply:
Appreciate the suggestion!